File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    
    int findBound(vector<int>& nums,int target,int lowBound){
        
        int n = nums.size();
        int l = 0;
        int r = n-1;
        
        while(l<=r){
            
            int mid =(l+r)/2;
            
            if(nums[mid]==target){
                if(lowBound==true){

                    if(mid==l or nums[mid-1]!=target){
                        return mid;
                    }

                    // when still on left side we have target element, we will find lowerbound on left side
                    r = mid-1;

                }else{

                    // we are finding upperbound

                    if(mid==r or nums[mid+1]!=target){
                        return mid;
                    }

                    l = mid+1;


                }
             }
            
          else if(nums[mid]>target){ 
              r = mid-1;
           }
            
         else{
               l = mid+1;
           }
        }
        
        return -1;
        
        
    }
    
    
    vector<int> searchRange(vector<int>& nums, int target) {
        
        int first = findBound(nums,target,true);
        
        if(first==-1) return {-1,-1};
        
        int last = findBound(nums,target,false);
        
        return {first,last};
           
    }
};

File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    
    def findBound(self,nums,target,findLB):
        
        l = 0
        r = len(nums)-1
        
        while l<=r:
            
            mid = (l+r)//2
            
            if nums[mid]==target:
                
                #find lowerbound
                if findLB == True:
                    
                    if mid==l or nums[mid-1]!=target:
                        return mid
                    else:
                        r = mid-1 # as lower bound search in right
                
                else:
                    # find upper bound
                    
                    if mid==r or nums[mid+1]!=target:
                        return mid
                    else:
                        l = mid+1
                        
            elif(nums[mid]<target):
                l = mid+1
            else:
                r = mid-1
        
        return -1
        
    
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        
        lower_bound = self.findBound(nums,target,True)
        
        if lower_bound == -1:
            return [-1,-1]
        
        upper_bound = self.findBound(nums,target,False)
        
        return [lower_bound,upper_bound]
        